src/lib/cmdline.c

```diff
@@ -123,49 +123,155 @@ char *cbm_parse_cmdline_file(const char *file)
         return strdup(buf);
 }
 
+/**
+ * Determine if there is a version of this file within the /etc/ tree that
+ * is "masking" the vendor file (i.e. it has the same name)
+ */
+static bool cbm_cmdline_disabled_by_mask(const char *root, char *path)
+{
+        autofree(char) *cmp = NULL;
+        autofree(char) *alt_path = NULL;
+        char *bcp = NULL;
+
+        cmp = strdup(path);
+        if (!cmp) {
+                return false;
+        }
+        bcp = basename(cmp);
+        if (!bcp) {
+                return false;
+        }
+
+        /* Find this path within /etc/ tree */
+        alt_path = string_printf("%s/%s/cmdline.d/%s", root, KERNEL_CONF_DIRECTORY, bcp);
+        return nc_file_exists(alt_path);
+}
+
+/**
+ * Determine if the file is disabled through a link to /dev/null and should be
+ * ignored. Together with @cbm_cmdline_disabled_by_mask, it is possible to disable
+ * a vendor file by masking the file into /etc/kernel/cmdline.d and linking it to
+ * /dev/null.
+ */
+static bool cbm_cmdline_disabled_by_link(char *path)
+{
+        autofree(char) *p = NULL;
+
+        /* GCC incorrectly complains about us freeing the return from realpath()
+         * which is allocated, however GCC believes it is heap storage.
+         */
+        p = realpath(path, NULL);
+        if (!p) {
+                return false;
+        }
+        return streq(p, "/dev/null");
+}
+
+/**
+ * Glob *.conf files within the given glob, and merge the resulting command
+ * line into the final stream
+ *
+ * @Returns negative code if the call failed, or the number of files processed.
+ */
+static int cbm_parse_cmdline_files_directory(const char *root, bool bump_start, bool check_masked,
+                                             char *globfile, FILE *memstr)
+{
+        glob_t glo = { 0 };
+        glo.gl_offs = 0;
+        glob(globfile, GLOB_DOOFFS, NULL, &glo);
+        int ret = -1;
+        size_t true_index = 0;
+
+        for (size_t i = 0; i < glo.gl_pathc; i++) {
+                char *argv = glo.gl_pathv[i];
+
+                /* If we're in a maskable directory, check if it's masked. */
+                if (check_masked && cbm_cmdline_disabled_by_mask(root, argv)) {
+                        LOG_DEBUG("Skipping masked file: %s", argv);
+                        continue;
+                }
+
+                /* If we're not in a maskable, check if it links to /dev/null */
+                if (!check_masked && cbm_cmdline_disabled_by_link(argv)) {
+                        LOG_DEBUG("Skipping disabled cmdline: %s", argv);
+                        continue;
+                }
+
+                /* Prevent accidental spaces for masked files */
+                true_index = i;
+
+                if (true_index > 0 || bump_start) {
```

Comment:
Maybe missing something but could you just use i here? i is just being assigned to true_index before true_index is used anyway and true_index doesn't appear to ever not match i when it is read.

Reply:
true_index might still be 0 while i > 0 because we might've skipped masked files, and this retains the "if I'm the first thing being printed add a space" logic

Reply:
Ah I see the purpse but this seems to be just a bug? Need true_index to get incremented at the end of the loop and not set to i in order to not insert a space when an earlier file was masked or set to /dev/null.

Reply:
..? Masked has a continue in the loop before we ever write the file

Reply:
Ah I see what you're saying, it should be ++'d not ='d.